shared/src/pool_fetching.rs

```diff
@@ -299,57 +315,81 @@ impl PoolFetching for PoolFetcher {
-
         batch.execute_all(MAX_BATCH_SIZE).await;
 
-        let mut results = Vec::with_capacity(futures.len());
-        for (pair, get_reserves, token0_balance, token1_balance) in futures {
-            let reserves = get_reserves.await;
-            let token0_balance = token0_balance.await;
-            let token1_balance = token1_balance.await;
-            if let (Ok(reserves), Ok(token0_balance), Ok(token1_balance)) =
-                (reserves, token0_balance, token1_balance)
-            {
-                // Some ERC20s (e.g. AMPL) have an elastic supply and can thus reduce the balance of their owners without any transfer or other interaction ("rebase").
-                // Such behavior can implicitly change the *k* in the pool's constant product formula. E.g. a pool with 10 USDC and 10 AMPL has k = 100. After a negative
-                // rebase the pool's AMPL balance may reduce to 9, thus k should be implicitly updated to 90 (figuratively speaking the pool is undercollateralized).
-                // Uniswap pools however only update their reserves upon swaps. Such an "out of sync" pool has numerical issues when computing the right clearing price.
-                // Note, that a positive rebase is not problematic as k would increase in this case giving the pool excess in the elastic token (an arbitrageur could
-                // benefit by withdrawing the excess from the pool without selling anything).
-                // We therefore exclude all pools where the pool's token balance of either token in the pair is less than the cached reserve.
-                if U256::from(reserves.0) <= token0_balance
-                    && U256::from(reserves.1) <= token1_balance
-                {
-                    results.push(Pool::uniswap(pair, (reserves.0, reserves.1)));
-                }
-            }
+        let mut results = Vec::new();
+        for future in futures {
+            results.push(future.await);
         }
-        results
+        handle_results(results)
+    }
+}
+
+struct FetchedPool {
+    pair: TokenPair,
+    reserves: Result<(u128, u128, u32), MethodError>,
+    token0_balance: Result<U256, MethodError>,
+    token1_balance: Result<U256, MethodError>,
+}
+
+fn handle_results(results: Vec<FetchedPool>) -> Result<Vec<Pool>> {
+    // Helper for the fold below.
+    // If the result is a node error then return it.
+    // If it is a contract error then skip this pool.
+    macro_rules! forward_error_or_skip {
+        ($result:expr, $acc:ident) => {
+            match $result {
+                Ok(t) => t,
+                Err(err) => match EthcontractErrorType::classify(&err) {
+                    EthcontractErrorType::Node => return Err(err.into()),
+                    EthcontractErrorType::Contract => return Ok($acc),
+                },
+            }
+        };
     }
+
+    results.into_iter().try_fold(Vec::new(), |mut acc, pool| {
+        let reserves = forward_error_or_skip!(pool.reserves, acc);
```

Comment:
Agree this is a little bit harder to understand. The reason in case it wasn't clear is that I didn't want to have the same fairly big match statement 3 times in the fold. And I needed a macro instead of a function so that it could do the return and continue directly.
I am also not sure how other people feel about it. Will try your suggestion.

Reply:
Changed it to use a normal function and 3 matches. I don't think the matches are too bad so I like this compromise.

Reply:
I liked the macro :stuck_out_tongue:. It brought up a bunch of nostalgia from the old `try!()` macro.

Reply:
@nlordell starring in: "Tell me you are a rust OG without telling me you are a rust OG"

Reply:
:rofl: